LAMMPS updates

rbberger opened the pull request:
- adds new stable release update
- adds latest patch release
- adds the missing download of a large potential file when using `+mesont`
- fix LAMMPS Python package installation for current releases. It now uses the correct Spack Python installation, installs itself via `pip`.

rbberger committed:
lammps: add new stable version 20230802.1

rbberger committed:
lammps: add missing potential download for +mesont

rbberger committed:
lammps: fix python package install

adamjstewart commented on var/spack/repos/builtin/packages/lammps/package.py:
```diff
@@ -343,6 +349,7 @@ class Lammps(CMakePackage, CudaPackage, ROCmPackage):
     )
```
Did you want to remove all these deprecated versions while you're at it?

rbberger replied:
LAMMPS is one of those codes where having the capability of installing an older version for reproducibility can be of interest. I have yet to find a good metric to determine when I should discard a specific LAMMPS version.

adamjstewart replied:
Usually the only reason to deprecate something is so it can be removed. You could also un-deprecate them.

rbberger replied:
Un-deprecation would defeat the purpose since we don't actually want people using those versions anymore unless they know what they are doing.

adamjstewart commented on var/spack/repos/builtin/packages/lammps/package.py:
```diff
@@ -604,6 +611,9 @@ def url_for_version(self, version):
+    depends_on("py-build", when="+python", type="build")
     depends_on("py-numpy", when="+python")
     depends_on("py-mpi4py", when="+python+mpi")
     depends_on("py-setuptools", when="@20220217:+python", type="build")
```
setuptools 42+ required now

adamjstewart replied:
Setuptools is also required at run-time due to a pkg_resources import

rbberger replied:
Added as run dependency and added `@42:` version specifier.

adamjstewart commented on var/spack/repos/builtin/packages/lammps/package.py:
```diff
@@ -604,6 +611,9 @@ def url_for_version(self, version):
+    depends_on("py-pip", when="+python", type="build")
+    depends_on("py-wheel", when="+python", type="build")
+    depends_on("py-build", when="+python", type="build")
     depends_on("py-numpy", when="+python")
     depends_on("py-mpi4py", when="+python+mpi")
```
Run-time deps aren't documented but I also see imports of matplotlib and pymol. Currently these are listed as build/link deps, which means the package isn't added to PYTHONPATH when loaded. Should these be build/run or build/link/run?

rbberger replied:
Those are dependencies to examples that aren't reachable from the Python package itself. Only the `lammps` package is installed.

adamjstewart replied:
> Currently these are listed as build/link deps, which means the package isn't added to PYTHONPATH when loaded. Should these be build/run or build/link/run?

Any thoughts on this?

rbberger replied:
Updated to be build/run dependencies.

adamjstewart commented on var/spack/repos/builtin/packages/lammps/package.py:
```diff
@@ -604,6 +611,9 @@ def url_for_version(self, version):
     depends_on("eigen@3:", when="+machdyn")
     depends_on("py-cython", when="+mliap+python")
     depends_on("py-cython", when="+ml-iap+python")
```
Cython should probably be a build-dep, there are no libs to link to

rbberger replied:
Limited to `build` dependency.

adamjstewart commented on var/spack/repos/builtin/packages/lammps/package.py:
```diff
@@ -835,3 +860,14 @@ def setup_run_environment(self, env):
+    @run_after("install")
+    def install_python(self):
+        # do LAMMPS Python package installation using pip
+        if self.spec.satisfies("@20230328: +python"):
```
So the Python variant exists for all versions but you only run this for new versions. Was this automatically installed in the past or something?

rbberger replied:
In the past it was part of the `install` target, but that was removed in favor of using pip directly. 

For the majority of users who build LAMMPS directly via CMake or Make, there is a `install-python` target that installs the Python package along with the LAMMPS shared library in the user's HOME. However, that mechanism isn't meant for packaging. Instead it should be treated as a regular Python package and installed via pip.

adamjstewart commented on var/spack/repos/builtin/packages/lammps/inject-modpath-for-python.patch:
Can you explain more about what this patch does? I don't love hard-coding things for Spack.

rbberger replied:
Without this the lammps shared library won't be found, since LD_LIBRARY_PATH won't be set.

adamjstewart replied:
We can set it if we need to.

rbberger replied:
If you look at the package.py, I already tried to do that, but that doesn't seem to work.

A simple test is to install `lammps+python+lib` into an environment and then do the following in a Python interpreter:

```python
import lammps
L = lammps.lammps()
```

Not sure what I'm missing to force it to be set.

rbberger replied:
@adamjstewart just tried this again on my personal system without the patch, and it behaves as expected. Not sure why it didn't work the other day. So I'll remove the patch and also add the missing DYLD_LIBRARY_PATH for MacOS.

adamjstewart reviewed (changes requested):
Only reviewed the Python bits

adamjstewart commented on var/spack/repos/builtin/packages/lammps/package.py:
```diff
@@ -833,5 +857,20 @@ def setup_build_environment(self, env):
-            env.prepend_path("LD_LIBRARY_PATH", self.prefix.lib)
-            env.prepend_path("LD_LIBRARY_PATH", self.prefix.lib64)
+            if self.spec.platform == "darwin":
+                env.prepend_path("DYLD_LIBRARY_PATH", self.prefix.lib)
+                env.prepend_path("DYLD_LIBRARY_PATH", self.prefix.lib64)
```
```suggestion
                env.prepend_path("DYLD_FALLBACK_LIBRARY_PATH", self.prefix.lib)
                env.prepend_path("DYLD_FALLBACK_LIBRARY_PATH", self.prefix.lib64)
```
This is only used if the libs can't be found by normal means.

rbberger committed:
Update var/spack/repos/builtin/packages/lammps/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

rbberger committed:
lammps: py-numpy and py-mpi4py should be build and run deps

rbberger committed:
lammps: add new 20231121 release

- MPIIO package has been removed -> disable mpiio variant
- LAMMPS_EXCEPTIONS is now always on -> disable exceptions variant
- CMake 3.16+ is now required
- Kokkos 4.1.0 is now supported
